documentation/src/main/asciidoc/user_guide/chapter-63-Streams.adoc

```diff
@@ -0,0 +1,416 @@
+link:http://docs.jboss.org/infinispan/{infinispanversion}/apidocs/org/infinispan/stream/CacheAware.html[CacheAware]
+interface the `injectCache` method be invoked before the accept method from the `Consumer` interface.
+
+There are plans to possibly adding a new forEach that instead takes a
+link:https://docs.oracle.com/javase/8/docs/api/java/util/function/BiConsumer.html[BiConsumer]
+which would allow for much cleaner API usage since a lambda could be used.
+
+==== Distributed Stream execution
+
+Distributed streams execution works in a fashion very similiar to map reduce.  Except in this case we are sending zero to many intermediate operations
+(map, filter etc.) and a single terminal operation to the various nodes.  The operation basically comes down to the following:
+
+. The desired segments are grouped by which node is the primary owner of the given segment
+
+. A request is generated to send to each remote node that contains the intermediate and terminal operations including which segments it should process
+
+.. The thread the terminal operation was invoked upon will perform the local operation directly
+
+.. Each remote node will receive this request and run the operations on an Infinispan remote thread and subsequently send the response back
+
+. The user thread will then gather the local response and remote responses together performing any kind of reduction required by the operations themselves.
+
+. Final reduced response is then returned to the user
+
+In most cases all operations are fully distributed, as in the operations are all fully applied on each remote node and usually only the last operation or something related may be
+reapplied to reduce the results from multiple nodes.  One important note is that intermediate values do not actually have to be serializable, it is the last value
+sent back that is the part desired (exceptions for various operations will be highlighted below).
+
+===== Terminal operator distributed result reductions
+The following paragraphs describe how the distributed reductions work for the various terminal operators.  Some of these are special in that an intermediate value may
+be required to be serializable instead of the final result.
+
+allMatch noneMatch anyMatch::
+The link:https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#allMatch-java.util.function.Predicate-[allMatch] 
+is ran on each node and then all the results are logically anded together locally to get the appropriate value.  The 
+link:https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#noneMatch-java.util.function.Predicate-[noneMatch] and 
+link:https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#anyMatch-java.util.function.Predicate-[anyMatch] perform logical or instead.  Note
+these methods also have early termination support in that if it returns early due to its normal stream operation it will also cause the distributed version to complete
+early as well.
```

Comment:
How about "These methods also have early termination support, stopping remote operations once the final result is known."